src/js/jsx/tools/ArtboardAdderOverlay.jsx

```diff
@@ -0,0 +1,305 @@
+ * all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+
+define(function (require, exports, module) {
+    "use strict";
+
+    var React = require("react"),
+        ReactDOM = require("react-dom"),
+        Fluxxor = require("fluxxor"),
+        FluxMixin = Fluxxor.FluxMixin(React),
+        StoreWatchMixin = Fluxxor.StoreWatchMixin,
+        Immutable = require("immutable"),
+        d3 = require("d3"),
+        _ = require("lodash");
+
+    var collection = require("js/util/collection");
+
+    // Used for debouncing the overlay drawing
+    var DEBOUNCE_DELAY = 200;
+
+    var ArtboardAdderOverlay = React.createClass({
+        mixins: [FluxMixin, StoreWatchMixin("tool", "document", "application", "ui", "panel")],
+
+        /**
+         * Owner group for all the overlay svg elements
+         *
+         * @type {SVGElement}
+         */
+        _scrimGroup: null,
+
+        /**
+         * UI Scale for drawing strokes visible at all zoom levels
+         *
+         * @type {number}
+         */
+        _scale: null,
+
+        /**
+         * During certain drags, especially with Artboards, we get a lot of
+         * UI updating events, document resizes etc. we debounce drawing functions
+         * so our overlays don't play catch up
+         *
+         * @type {function}
+         */
+        _drawDebounced: null,
+
+        getStateFromFlux: function () {
+            var flux = this.getFlux(),
+                applicationStore = flux.store("application"),
+                toolStore = flux.store("tool"),
+                modalState = toolStore.getModalToolState(),
+                currentDocument = applicationStore.getCurrentDocument(),
+                vectorMaskMode = toolStore.getVectorMode();
+
+            return {
+                document: currentDocument,
+                modalState: modalState,
+                vectorMaskMode: vectorMaskMode
+            };
+        },
+
+        shouldComponentUpdate: function (nextProps, nextState) {
+            // We check for the cheaper flags before plucking layers from document models
+            if (this.state.vectorMaskMode !== nextState.vectorMaskMode ||
+                this.state.modalState !== nextState.modalState ||
+                this.props.transformString !== nextProps.transformString) {
+                return true;
+            }
+
+            var lastLayers = this.state.document.layers.selected,
```

Comment:
I wonder if this can go further by saying something like: return false if there were no artboards selected in either the next or previous state.